pkg/controllers/manila/manila.go

```diff
@@ -0,0 +1,221 @@
+	csiControllers     []Runnable
+	controllersRunning bool
+	eventRecorder      events.Recorder
+}
+
+type Runnable interface {
+	Run(ctx context.Context, workers int)
+}
+
+const (
+	// Minimal interval between controller resyncs. The controller will detect
+	// new share types in Manila and create StorageClasses for them at least
+	// once per this interval.
+	resyncInterval = 1 * time.Minute
+
+	operatorConditionPrefix = "ManilaController"
+)
+
+func NewController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	informers v1helpers.KubeInformersForNamespaces,
+	openStackClient *openStackClient,
+	csiControllers []Runnable,
+	eventRecorder events.Recorder) factory.Controller {
+
+	scInformer := informers.InformersFor("").Storage().V1().StorageClasses()
+	c := &Controller{
+		operatorClient:     operatorClient,
+		kubeClient:         kubeClient,
+		storageClassLister: scInformer.Lister(),
+		openStackClient:    openStackClient,
+		csiControllers:     csiControllers,
+		eventRecorder:      eventRecorder,
+	}
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(operatorClient).ResyncEvery(resyncInterval).WithInformers(
+		operatorClient.Informer(),
+		scInformer.Informer(),
+	).ToController("ManilaController", eventRecorder)
+}
+
+func (c *Controller) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("Manila sync started")
+	defer klog.V(4).Infof("Manila sync finished")
+
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorv1.Managed {
+		return nil
+	}
+
+	shareTypes, err := c.openStackClient.GetShareTypes()
+	if err != nil {
+		switch err.(type) {
+		case *gophercloud.ErrEndpointNotFound:
+			// OpenStack does not support manila, report the operator as disabled
+			return c.setDisabled("This OpenStack does not provide Manila service")
+		default:
+			return err
+		}
+	}
+
+	if len(shareTypes) == 0 {
+		klog.V(4).Infof("Manila does not provide any share types")
+		return c.setDisabled("Manila does not provide any share types")
+	}
+	// Manila has some shares: start the actual CSI driver controller sets
+	if !c.controllersRunning {
+		klog.V(4).Infof("Starting CSI driver controllers")
+		for _, ctrl := range c.csiControllers {
+			go ctrl.Run(ctx, 1)
+		}
+		c.controllersRunning = true
+	}
+	err = c.syncStorageClasses(ctx, shareTypes)
+	if err != nil {
+		return err
+	}
+
+	return c.setEnabled()
+}
+
+func (c *Controller) syncStorageClasses(ctx context.Context, shareTypes []sharetypes.ShareType) error {
+	var errs []error
+	for _, shareType := range shareTypes {
+		klog.V(4).Infof("Syncing storage class for shareType type %s", shareType.Name)
+		sc := c.generateStorageClass(shareType)
+		err := c.applyStorageClass(ctx, sc)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) != 0 {
+		return errors.NewAggregate(errs)
+	}
+	return nil
```

Comment:
I think you can just `return errors.NewAggregate(errs)`

Reply:
Fixed